torch/optim/adam.py

```diff
@@ -429,13 +436,43 @@ def _single_tensor_adam(
         # Decay the first and second moment running average coefficient
         exp_avg.lerp_(grad, 1 - device_beta1)
 
-        exp_avg_sq.mul_(beta2).addcmul_(grad, grad.conj(), value=1 - beta2)
+        # Nested if is necessary to bypass jitscript rules
+        if differentiable and isinstance(beta2, Tensor):
+            if beta2.requires_grad:
+                # Using lerp to only use 2 operations bc addcmul's value cannot be a tensor
+                # Showing equivalence of differentiable path and nondifferentiable path
+                # expavg * b2 + grad^2 * (1-b2)
+                #           add expavg * (1-b2) - expavg * (1-b2) = 0
+                # expavg * b2 + expavg * (1-b2) - expavg * (1-b2) + grad^2 * (1-b2)
+                # expavg - expavg * (1-b2) + grad^2 * (1-b2)
+                # expavg + (grad^2 - expavg) * (1-b2)
+                # expavg.lerp(grad^2, 1-beta2)
+                exp_avg_sq.lerp_(grad * grad.conj(), weight=1 - beta2)
```

Comment:
- btw wondering if grad was already converted to real in the code `if torch.is_complex(param): grad = torch.view_as_real(grad)` above, is it worth to instead do `torch.square(grad)` and if it would do fewer of memory loads because it knows that it just has a single tensor as input? (i.e. `torch.mul(a, a)` versus `torch.square(a)`) 
- And if temp copy for the `grad * grad.conj()` is not the problem, does it make sense to unite all three branches?

Reply:
square makes sense to me! I think it might be possible to unite the branches, but since every other differentiable path I've written so far has this branched structure, it makes senseto me to leave the three branches for clarity, but I'd love to know what Jane thinks! I dont want to bother her since it's holiday time so I'd love to know what she thinks when it's not holiday time.

Reply:
using square/getting rid of grad.conj() for this branch is fine according to @vadimkantorov's point!

We shouldn't unite the branches though, because in 

`exp_avg_sq.mul_(beta2).addcmul_(grad, grad.conj(), value=1 - beta2)`

the grad.conj() is a view (not materialized into a Tensor) and so we don't want to introduce a materialized tensor in those branches

Reply:
If grad is converted to real, should everywhere `grad.conj()` be replaced by just `grad` for clarity that no actual conjugation is done?